File: evaluation/framework/evidence.py
"""Small helpers for reading evaluation evidence bundles."""

from __future__ import annotations

from collections.abc import Iterable


def snapshot_names(spec: dict) -> tuple[str, ...]:
    snapshot = spec.get("snapshot", "after")
    if snapshot == "both":
        return ("before", "after")
    if snapshot in {"before", "after"}:
        return (snapshot,)
    raise ValueError("snapshot must be one of: before, after, both")


def entities_for_snapshot(evidence: dict, snapshot: str) -> dict:
    section = (evidence or {}).get(snapshot)
    if not isinstance(section, dict):
        return {}
    entities = section.get("entities")
    if not isinstance(entities, dict):
        return {}
    return entities


def entity_for_snapshot(evidence: dict, snapshot: str, entity_id: str) -> dict | None:
    entity = entities_for_snapshot(evidence, snapshot).get(entity_id)
    return entity if isinstance(entity, dict) else None


def missing_snapshots(evidence: dict, entity_id: str, snapshots: Iterable[str]) -> list[str]:
    missing: list[str] = []
    for snapshot in snapshots:
        if entity_for_snapshot(evidence, snapshot, entity_id) is None:
            missing.append(snapshot)
    return missing

File: packages/eval/src/evaluation/evidence.ts
/** Helpers for reading evaluation evidence bundles. */

export function snapshotNames(spec: Record<string, any>): string[] {
  const snapshot = spec.snapshot ?? "after";
  if (snapshot === "both") return ["before", "after"];
  if (snapshot === "before" || snapshot === "after") return [snapshot];
  throw new Error("snapshot must be one of: before, after, both");
}

export function entitiesForSnapshot(evidence: Record<string, any>, snapshot: string): Record<string, any> {
  const section = (evidence ?? {})[snapshot];
  if (section === null || typeof section !== "object" || Array.isArray(section)) return {};
  const entities = section.entities;
  if (entities === null || typeof entities !== "object" || Array.isArray(entities)) return {};
  return entities;
}

export function entityForSnapshot(
  evidence: Record<string, any>,
  snapshot: string,
  entityId: string,
): Record<string, any> | null {
  const entity = entitiesForSnapshot(evidence, snapshot)[entityId];
  return entity !== null && typeof entity === "object" && !Array.isArray(entity) ? entity : null;
}

export function missingSnapshots(evidence: Record<string, any>, entityId: string, snapshots: string[]): string[] {
  return snapshots.filter((snapshot) => entityForSnapshot(evidence, snapshot, entityId) === null);
}
